File: 44_JSON読み込み/python/main.py
#!/usr/bin/env python
# coding: -utf8

import user_input as ui
import sys
import json


def validate_not_empty(s):
    stripped = s.strip()
    if len(stripped) == 0:
        raise ValueError

    return stripped


def find_first(iterable, pred):
    default = False
    return next(filter(pred, iterable), default)


def disp(product):
    print("Name: %s" % product['name'])
    print("Price: $%.2f" % product['price'])
    print("Quantity on hand: %d" % product['quantity'])


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <json file>", file=sys.stderr)
        sys.exit(1)

    products = None
    with open(sys.argv[1], 'rt') as f:
        loaded = json.load(f)
        products = loaded["products"]

    WRONG_INPUT_MSG = "Sorry, that product was not found in our inventory."

    while True:
        name = ui.read_user_input_with_validator(
            "What is the procuct name? ", validate_not_empty, WRONG_INPUT_MSG)

        def pred(procuct): return procuct['name'] == name

        found = find_first(products, pred)

        if found:
            disp(found)
            break

        print(WRONG_INPUT_MSG)

File: 44_JSON読み込み/cpp/main.cc
#include <fstream>
#include <iomanip>
#include <iostream>

#include "product.h"
#include "strings.h"
#include "user_input.h"

using namespace product;

namespace {

bool validate_not_empty(const std::string& s, std::string& value) {
    auto trimed_s = strings::trim(s);
    if (trimed_s.empty()) {
        return false;
    }
    value = trimed_s;
    return true;
}

}  // namespace

int main(int argc, char const* argv[]) {
    std::vector<Product> products;
    if (argc < 2) {
        std::cerr << "Usage: " << *argv << " <json file>" << std::endl;
        exit(EXIT_FAILURE);
    }

    std::ifstream ifs(*++argv);
    bool success = product::parse(ifs, products);
    ifs.close();
    if (!success) {
        exit(EXIT_FAILURE);
    }

    const std::string not_found_msg =
        "Sorry, that product was not found in our inventory.";

    for (;;) {
        auto name = read_user_input_with_validator<std::string>(
            "What is the procuct name? ", validate_not_empty, not_found_msg);

        auto pred = [name](const Product& p) { return p.name() == name; };

        auto it = std::find_if(std::begin(products), std::end(products), pred);
        if (it != std::end(products)) {
            // found
            it->disp();
            break;
        }
        std::cout << not_found_msg << std::endl;
    }
    exit(EXIT_SUCCESS);
}
